### DISCUSSION TITLE
Use ChatCompletionRequest for Chat inputs

### DISCUSSION THREAD

**Discussion by Shiti** [rustic-ai/ui-components#310] [Ideas]
Most LLM Agent libraries and services now conform to a standard interface similar to [OpenAI's ChatCompletionRequest](https://platform.openai.com/docs/api-reference/chat). It would make [ui-components](https://github.com/rustic-ai/ui-components) library easier to use and adopt if the input components send messages in this format instead of custom formats such as `text` and `filesWithText`.

OpenAI only documents using image and audio for content types of [ChatCompletionRequest](https://platform.openai.com/docs/api-reference/chat) but we could make it more generic to support any file type. 

For example, the interface could be something like,

```typescript
interface Content {
  type: string,
  text?: string,
  url?: string,
}

export interface ChatCompletionRequest {
  messages: { content: Content[], role: string }[]
}
```

and some sample corresponding messages sent on the websocket could be

```json
{
  "id": "4f6f6c72-4328-4170-8d35-7fabe322be3e",
  "timestamp": "2024-11-29T18:03:25.120Z",
  "sender": {
    "id": "46ec2629",
    "name": "Anonymous"
  },
  "conversationId": "7fabe322be3e",
  "format": "ChatCompletionRequest",
  "data": {
    "messages": [
      {
        "content": [
          {
            "type": "text",
            "text": "what was the monthly average sales reported for the last quarter?"
          },
          {
            "type": "file",
            "url": "https://somefile.xls"
          }
        ],
        "role": "user"
      }
    ]
  }
}
```

**Pros**
* Simply integration experience for developers using `ui-components` library with LLM Agents

**Cons**
* This would break existing applications

**Comment by lyjeileen** [comment]
- Currently, the message is sent without waiting for the URL to be returned. Should we change this behavior?
- Just to confirm, are the sender, id, and timestamp attributes specific to the last message in the `messages.content` array? Will the content continue to grow as the chat progresses?

**Reply by Shiti** [reply]
> * Currently, the message is sent without waiting for the URL to be returned. Should we change this behavior?

If thats the case, yes we may need to update that

> * Just to confirm, are the sender, id, and timestamp attributes specific to the last message in the `messages.content` array? Will the content continue to grow as the chat progresses?

The example I shared above is a single ui-components `Message` sent when someone clicks the send button on the text-input or multimodal-input components. 

No, I dont think we should append all previous messages to the content. That is something backend should handle if desired.

**Reply by lyjeileen** [reply]
- Could you add a comparison of message types for before and after? 
- I thought it was possible to have both url and text in a single type. Could you create two separate types for Content?

**Comment by Shiti** [comment]
Following are some examples of current messages sent vs future ones with this change - 
_Note:_ I have skipped the other fields in ui-components message which will not be effected.

| Current  Implementation                                                                                                                                                                                     | Proposed Changes                                                                                                                                                                                                                                                                                                                                                                                                              |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| <pre>{<br/>&nbsp;&nbsp;"format":&nbsp;"text",<br/>&nbsp;&nbsp;"data":&nbsp;{<br/>    "text":&nbsp;"breakdown&nbsp;what&nbsp;llms&nbsp;are&nbsp;for&nbsp;a&nbsp;non-techie"<br/>&nbsp;&nbsp;}<br/>}</pre>    | <pre>{<br/>  "format":&nbsp;"ChatCompletionRequest",<br/>  "data":&nbsp;{<br/>    "messages":&nbsp;[<br/>      {<br/>        "content":&nbsp;[<br/>          {<br/>            "type":&nbsp;"text",<br/>            "text":&nbsp;"breakdown&nbsp;what&nbsp;llms&nbsp;are&nbsp;for&nbsp;a&nbsp;non-techie"<br/>&nbsp;&nbsp;         }<br/>        ],<br/>        "role":&nbsp;"user"<br/>      }<br/>    ]<br/>  }<br/>}</pre> |
| <pre>{<br/>&nbsp;&nbsp;"format":&nbsp;"filesWithText",<br/>&nbsp;&nbsp;"data":&nbsp;{<br/>    "files":[<br/>      {<br/>        "name": "somefile.xls" <br/>      }<br/>    ]<br/>&nbsp;&nbsp;}<br/>}</pre> | <pre>{<br/>  "format":&nbsp;"ChatCompletionRequest",<br/>  "data":&nbsp;{<br/>    "messages":&nbsp;[<br/>      {<br/>        "content":&nbsp;[<br/>          {<br/>            "type":&nbsp;"file",<br/>            "url":&nbsp;"https://somefile.xls"<br/>&nbsp;&nbsp;         }<br/>        ],<br/>        "role":&nbsp;"user"<br/>      }<br/>    ]<br/>  }<br/>}</pre>  |
| <pre>{<br/>&nbsp;&nbsp;"format":&nbsp;"filesWithText",<br/>&nbsp;&nbsp;"data":&nbsp;{<br/>    "text":&nbsp;"what was the monthly average sales reported for the last quarter?"<br/>    "files":[<br/>      {<br/>        "name": "somefile.xls" <br/>      }<br/>    ]<br/>&nbsp;&nbsp;}<br/>}</pre>    | <pre>{<br/>  "format":&nbsp;"ChatCompletionRequest",<br/>  "data":&nbsp;{<br/>    "messages":&nbsp;[<br/>      {<br/>        "content":&nbsp;[<br/>          {<br/>            "type":&nbsp;"text",<br/>            "text":&nbsp;"what was the monthly average sales reported for the last quarter?"<br/>&nbsp;&nbsp;         }<br/><br/>          {<br/>            "type":&nbsp;"file",<br/>            "url":&nbsp;"https://somefile.xls"<br/>&nbsp;&nbsp;         }<br/>        ],<br/>        "role":&nbsp;"user"<br/>      }<br/>    ]<br/>  }<br/>}</pre> |



> I thought it was possible to have both url and text in a single type. Could you create two separate types for Content?

@lyjeileen yes, I agree, it would be simpler to define two types for the content with clear indication of the fields.

**Reply by lyjeileen** [reply]
This makes sense. Does this also mean we are going to replace `Text` and `Multipart` components with `ChatCompletionRequest` component?

**Reply by Shiti** [reply]
I was thinking of updating `Multipart` and leaving `Text` as is.  Or I could avoid changing those but would need to have something else for displaying `ChatCompletionRequest` . What do you think ?

**Reply by lyjeileen** [reply]
Do you think `Multipart` would still be useful to others? If the backend normally sends messages in the `ChatCompletionRequest` format, it makes sense to update `Multipart`.

**Comment by Shiti** [comment]
We decided to keep the Multipart and introduced a separate component for ChatCompletion. The changes discussed were released in [v.0.0.34](https://github.com/rustic-ai/ui-components/releases/tag/v.0.0.34)
